### PR TITLE
Bug 1702206: Refactor Uploader into class for external use

### PR DESCRIPTION
**Pull request by ChinYing-Li**
### Pull Request checklist ###

- [x] **Quality**: Make sure this PR builds and runs cleanly.
  - Inside the `glean/` folder, run:
    - `npm run test` Runs _all_ tests
    - `npm run lint && npm run lint:circular-deps` Runs _all_ linters
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes a changelog entry to `CHANGELOG.md` or an explanation of why it does not need one
- [x] **Documentation**: This PR includes documentation changes, an explanation of why it does not need that or a follow-up bug has been filed to do that work
  - Documentation should be added to [The Glean Book](https://mozilla.github.io/glean/book/index.html) when making changes to Glean's user facing API
    - When changes to the Glean Book are necessary, link to the corresponding PR on the [`mozilla/glean`](https://github.com/mozilla/glean) repository
  - Documentation should be added to [the Glean.js developer documentation](https://github.com/mozilla/glean.js/tree/main/docs) when making internal code changes

This PR aims to tackle bug 1702206; no new test added but some test cases have been modified.
Please ignore my question on Bugzilla, as I now understand why this approach is preferred. PR ready for review, ~~and will modify `CHANGELOG.md` if there's no problem~~. Thanks!

### CODE REVIEW

File: glean/src/core/upload/uploader.ts [hunk 1]
```diff
@@ -28,17 +28,22 @@ export const enum UploadResultStatus {
...
+  readonly status?: number;
   // The status of an upload attempt
-  result: UploadResultStatus
+  readonly result: UploadResultStatus;
+
+  constructor(result: UploadResultStatus, status?: number) {
+    this.result = result;
+    this.status = status;
+  }
 }
```
Comment (brizental) [inline glean/src/core/upload/uploader.ts:41]: You can  make this whole constructor declaration more succint, with the following syntax:

```ts
export class UploadResult {
   constructor(
      // The status of an upload attempt
      readonly result: UploadResultStatus,
      // The status is only present if `result` is UploadResultStatus.Success
      readonly status?: number
   ): {}
}
```

This is equivalient to:


```ts
export class UploadResult {
  // The status is only present if `result` is UploadResultStatus.Success
  readonly status?: number;
  // The status of an upload attempt
  readonly result: UploadResultStatus;

  constructor(result: UploadResultStatus, status?: number) {
    this.result = result;
    this.status = status;
  }
}
```

File: glean/tests/unit/platform/qt/uploader.spec.ts [hunk 2]
```diff
@@ -7,7 +7,7 @@ import assert from "assert";
 import sinon from "sinon";
 
 import QtUploader from "../../../../src/platform/qt/uploader";
-import { DEFAULT_UPLOAD_TIMEOUT_MS, UploadResultStatus } from "../../../../src/core/upload/uploader";
+import {DEFAULT_UPLOAD_TIMEOUT_MS, UploadResult, UploadResultStatus} from "../../../../src/core/upload/uploader";
```
Comment (brizental) [inline glean/tests/unit/platform/qt/uploader.spec.ts:10]: nit: Space before `}`

File: glean/tests/unit/platform/webext/uploader.spec.ts [hunk 3]
```diff
@@ -8,7 +8,7 @@ import sinon from "sinon";
 
 import BrowserUploader from "../../../../src/platform/webext/uploader";
 import type { JSONObject } from "../../../../src/core/utils";
-import { UploadResultStatus } from "../../../../src/core/upload/uploader";
+import {UploadResult, UploadResultStatus} from "../../../../src/core/upload/uploader";
```
Comment (brizental) [inline glean/tests/unit/platform/webext/uploader.spec.ts:11]: nit: Space after `{` and before `}`

File: samples/web-extension/typescript/package.json [hunk 4]
```diff
@@ -21,12 +21,12 @@
...
+    "strip-ansi": "^7.0.0",
     "ts-loader": "^9.2.3",
+    "ts-node": "^10.1.0",
     "typescript": "^4.3.5",
     "web-ext-types": "^3.2.1",
     "webpack": "^5.44.0",
-    "webpack-cli": "^4.7.2",
-    "ts-node": "^10.1.0",
-    "strip-ansi": "^7.0.0"
+    "webpack-cli": "^4.7.2"
```
Comment (brizental) [inline samples/web-extension/typescript/package.json:30]: Hm, why are these changes part of this PR?

File: CHANGELOG.md [hunk 5]
```diff
@@ -2,6 +2,8 @@
 
 [Full changelog](https://github.com/mozilla/glean.js/compare/v0.17.0...main)
 
+* [#534](https://github.com/mozilla/glean.js/pull/534): Refactor Uploader into abstract class for external use
```
Comment (brizental) [inline CHANGELOG.md:5]: The changelog is aimed at users, so it's ideal to write the message here tailored for these readers. They don't necessarily have context related to internals and mostly care about how changes affect external APIs.

How about, something more like:

> Expose `Uploader` base class through `@mozilla/glean/<platform>/uploader` entry point.

WDYT?

**Review by brizental — CHANGES_REQUESTED** [review]
Thanks for the work here @ChinYing-Li ! This looks great. I left some very minor comments, but otherwise this is good to go.

File: samples/web-extension/typescript/package.json (hunk 4 above)
Comment (ChinYing-Li) [reply samples/web-extension/typescript/package.json:30]: I ran the sample to test out the change, and somehow `package.json` got modified? I will revert the change.

**Review by brizental — APPROVED** [review]
Thanks for working on this @ChinYing-Li ! Looks great :D

**Pull request merged**